lib/util/gitfs/git/index.ts

```diff
@@ -0,0 +1,599 @@
+    'remote: Invalid username or password',
+    'gnutls_handshake() failed',
+    'The requested URL returned error: 5',
+    'The remote end hung up unexpectedly',
+    'access denied or repository not exported',
+    'Could not write new index file',
+    'Failed to connect to',
+    'Connection timed out',
+  ];
+  for (const errorStr of platformFailureStrings) {
+    if (err.message.includes(errorStr)) {
+      throw new ExternalHostError(err, 'git');
+    }
+  }
+}
+
+function localName(branchName: string): string {
+  return branchName.replace(/^origin\//, '');
+}
+
+function throwBaseBranchValidationError(branchName: string): never {
+  const error = new Error(CONFIG_VALIDATION);
+  error.validationError = 'baseBranch not found';
+  error.validationMessage =
+    'The following configured baseBranch could not be found: ' + branchName;
+  throw error;
+}
+
+async function isDirectory(dir: string): Promise<boolean> {
+  try {
+    return (await fs.stat(dir)).isDirectory();
+  } catch (err) {
+    return false;
+  }
+}
+
+async function getDefaultBranch(git: Git.SimpleGit): Promise<string> {
+  // see https://stackoverflow.com/a/44750379/1438522
+  try {
+    const res = await git.raw(['symbolic-ref', 'refs/remotes/origin/HEAD']);
+    return res.replace('refs/remotes/origin/', '').trim();
+  } catch (err) /* istanbul ignore next */ {
+    checkForPlatformFailure(err);
+    if (
+      err.message.startsWith(
+        'fatal: ref refs/remotes/origin/HEAD is not a symbolic ref'
+      )
+    ) {
+      throw new Error(REPOSITORY_EMPTY);
+    }
+    throw err;
+  }
+}
+
+let config: LocalConfig = {} as any;
+
+let git: Git.SimpleGit | undefined;
+
+let cwd: string | undefined;
+
+let privateKeySet = false;
+
+async function resetToBranch(branchName: string): Promise<void> {
+  logger.debug(`resetToBranch(${branchName})`);
+  await git.raw(['reset', '--hard']);
+  await git.checkout(branchName);
+  await git.raw(['reset', '--hard', 'origin/' + branchName]);
+  await git.raw(['clean', '-fd']);
+}
+
+async function deleteLocalBranch(branchName: string): Promise<void> {
+  await git.branch(['-D', branchName]);
+}
+
+async function cleanLocalBranches(): Promise<void> {
+  const existingBranches = (await git.raw(['branch']))
+    .split('\n')
+    .map((branch) => branch.trim())
+    .filter((branch) => branch.length)
+    .filter((branch) => !branch.startsWith('* '));
+  logger.debug({ existingBranches });
+  for (const branchName of existingBranches) {
+    await deleteLocalBranch(branchName);
+  }
+}
+
+export async function getSubmodules(): Promise<string[]> {
+  return (
+    (await git.raw([
+      'config',
+      '--file',
+      '.gitmodules',
+      '--get-regexp',
+      'path',
+    ])) || ''
+  )
+    .trim()
+    .split(/[\n\s]/)
+    .filter((_e: string, i: number) => i % 2);
+}
+
+export function isInitialized(): boolean {
+  return !!git;
+}
+
+export function cleanRepo(): void {
+  if (isInitialized()) {
+    // no-op
+  }
+}
```

Comment:
```suggestion
```

@rarkins Can we remove it, as it's a no-op? Or should we call git clean and checkout default branch?

Reply:
Unfortunately I forget the significance of this cleaning. Was it already a no-op before this PR?

Reply:
Yes, it was

Reply:
ok, let us know when ready for review again

Reply:
Given that (1) we'll do `getDefaultBranch` refactoring later and (2) we're leaving `cleanRepo()` no-op function (aren't we?), I think it's ready

Reply:
Also we can do both in later PR

Reply:
We can refactor this in a later pr